### DISCUSSION TITLE
How to smooth a surface?

### DISCUSSION THREAD

**Discussion by M8kmyday** [mfem/mfem#4559] [Q&A]
I have generated a 2D triangular mesh in 3D space, calculated scalar values at the mesh vertices, and set a GridFunction with those values using an H1 finite element space with order=1 and dim=2.  The results look like this:
![Screenshot from 2024-10-30 08-45-46](https://github.com/user-attachments/assets/ae8b71bf-c090-4cc5-90bc-ed5314ca9a1f)
I would like to smooth this plot to obtain a publication quality figure.  I can certainly make the plot smoother by increasing the mesh density, but this gets to be very slow for a smooth-looking result.  Is there a way within MFEM to apply smoothing to obtain a smooth plot from the coarser data?

**Comment by pazner** [comment]
It looks like you are using ParaView to do the visualization. ParaView can use a data array for the normal vectors. This can be used to make the surface look much smoother.

You can create a piecewise linear GridFunction whose values at the vertices are the averages of the normal vectors of the adjacent elements, and then use this array as the surface normal array in ParaView.

```c++
H1_FECollection h1_fec(1, bdr_mesh.Dimension());
FiniteElementSpace h1_fes(&bdr_mesh, &h1_fec, bdr_mesh.SpaceDimension());

GridFunction normals(&h1_fes);
normals = 0.0;

Array<int> multiplicity(normals.Size());
multiplicity = 0;

for (int e = 0; e < bdr_mesh.GetNE(); ++e)
{
   auto &T = *bdr_mesh.GetElementTransformation(e);
   Array<int> vdofs;
   h1_fes.GetElementVDofs(e, vdofs);
   const auto &ir = h1_fes.GetFE(e)->GetNodes();
   for (int i = 0; i < ir.Size(); ++i)
   {
      T.SetIntPoint(&ir[i]);
      Vector normal_vector(bdr_mesh.SpaceDimension());
      CalcOrtho(T.Jacobian(), normal_vector);
      for (int d = 0; d < bdr_mesh.SpaceDimension(); ++d)
      {
         const int vd = vdofs[i + d*ir.Size()];
         normals[vd] += normal_vector[d];
         ++multiplicity[vd];
      }
   }
}

for (int i = 0; i < normals.Size(); ++i)
{
   normals[i] /= multiplicity[i];
}

ParaViewDataCollection pv("SmoothedSurface", &bdr_mesh);
pv.SetPrefixPath("ParaView");
pv.RegisterField("normals", &normals);
pv.SetCycle(0);
pv.SetTime(0.0);
pv.Save();
```

Without normal array:

![1](https://github.com/user-attachments/assets/88eb04dd-cfff-4c5a-b826-497aa2d978c8)

With normal array:

![2](https://github.com/user-attachments/assets/ef08fc8f-9341-4430-bb27-afb46bd20161)

The setting is here:

![image](https://github.com/user-attachments/assets/fa7db101-498c-4f5c-9308-81b89756a663)

**Comment by M8kmyday** [comment]
Yes, that works quite well.  Here is the smoothed plot:
![Screenshot from 2024-10-30 14-21-38](https://github.com/user-attachments/assets/fa0d932f-e69d-46ba-a32c-d0bb16c6d116)
In case someone else gives this a try, my compiler was erroring out on the auto &T, so I just explicitly typed it as ElementTransformation *T.  Also, my ParaView version was 5.7.0 from the Ubuntu rpm package, which is too far out-of-date and does not have the Normal Array option for the Lighting, but downloading the latest version directly from www.paraview.org fixed that problem.

Thanks for the quick help!

Brian
